Add HK ADC (ADS1015) functionality on CUBES

sonals301 opened the pull request:
Add HK ADC reading for measurement of Voltage, Temperature and Current from CUBES

Commit:
Add hk_adc_init() at the start of main()

Commit:
first commit: add hk_adc init fn, reg_read, reg_write - working so far

Commit:
Added conversion reg read and config fns - tested but not worked yet

Commit:
WIP: refactoring of ADC config defs

Commit:
WIP: yayy! voltage reading works, refactored code

Commit:
add Battery current and Citi Temp read fns - mismatch in values as of now

Commit:
WIP: GPIO8 read instead of an i2c transaction after conversion

Commit:
WIP: working code - correct batt volt and curr readings, temp reading 30 deg as of now

Commit:
removed unwanted commented code

Commit:
added documentation for the fns

Commit:
added avg calc of last 10 samples, get fns for avg values for volt, curr and citi_temp

Commit:
added hk adc fn calls in exp_handler to be added into HK stream at the end, added fn calls in main only for testing purposes

Commit:
working code for REQ_HK with HK_ADC - check linker flag for use fo 'float' in project properties

Commit:
removed the 'float' property, storing as unsigned integer into hk_data

Commit:
removed ADC conversion equation to real values for batt v,I and citi temp; added documentation for some fns

Commit:
code cleanup - removed commented and unneeded code

Commit:
added CITI_TEMP in DAQ stream

Commit:
rebased with rebinning branch

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+/*
```
The usual: GPL header here please.

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+#define TX_LENGTH                       1u
+#define RX_LENGTH                       2u
+#define DUMMY_SERIAL_ADDR               0xFF
+#define NUM_SAMPLES_TO_AVG              10u
```
Good habit to put the macro assignment value into parantheses, e.g., `#define TX_LENGTH (1u)`.

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+    uint8_t err = HK_ADC_ERR_INIT_FAILED;
+
+    // By default, ADC is configured to point towards CITI_TEMP as MUX input
+    uint16_t config = HK_ADC_CONFIG_MUX_SINGLE_2 | HK_ADC_CONFIG_FSR_2 | HK_ADC_CONFIG_MODE_SINGLE_SHOT | HK_ADC_CONFIG_DR_1600SPS | HK_ADC_CONFIG_CMODE_TRAD | HK_ADC_CONFIG_CPOL_ACTIVE_LOW | HK_ADC_CONFIG_CLAT_NON_LATCH | HK_ADC_CONFIG_CQUE_DISABLE;
```
Two things about this line:

1. Please code lines within 80 columns, it makes it easier to read without needing to scroll right. (Yes, I know you would then need to scroll down, but that is more natural than scrolling right and left...)
2. Make `config` a `const uint16_t CONFIG` (capitalize it to reflect it is a constant). Seems more logical like this, given it is only used once and within this function.

sonals301 replied:
Good points! I made the change

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+    uint8_t send_buffer[2] = {config>>8, config};
+
+    // Set the clock for i2c0 instance
+    MSS_I2C_init(&g_mss_i2c0, DUMMY_SERIAL_ADDR, MSS_I2C_PCLK_DIV_256); //10^8/256 = 390,625Hz =~ 390KHz
```
Same as above with the 80-column limit. You could put the comment at the end of this line in the already commented line above and fit within 80 cols.

sonals301 replied:
Haha yeah, sorry! I don't remember why I even put it there in the first place. My bad! 
Made the change.

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+
+    if (err == HK_ADC_NO_ERR)
+    {
+        // Let the rest of the configurations remain intact, just change MUX input config and start single-shot conversion
```
80 cols

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+            err = hk_adc_reg_read(HK_ADC_REG_CONVERSION, &read_value);
+            if (err == HK_ADC_NO_ERR)
+            {
+                // Right shift the received result by 4 bits. Conversion value are 12 most significant bits
```
80 cols

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+ * value is returned in the unsigned 16-bit parameter passed to the function as
+ * argument.
+ *
+ * @param batt_volt[in]     pointer to unsigned 16-bit variable to store the result in.
```
This is actually an `[out]` param.

sonals301 replied:
True! Made the change.

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+/**
+ * @brief HK ADC read Battery Voltage function
+ *
+ * This function returns the raw battery voltage value as read by the ADC. ADC
```
I would rephrase "returns" here, to avoid confusion in the future.

sonals301 replied:
Made the change. Hope it looks better now.

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+/**
+ * @brief HK ADC read Battery Current function
+ *
+ * This function returns the raw battery current value as read by the ADC. ADC
```
I would rephrase "returns" here, to avoid confusion in the future.

sonals301 replied:
Same as above

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+ * value is returned in the unsigned 16-bit parameter passed to the function as
+ * argument.
+ *
+ * @param batt_curr[in]     pointer to unsigned 16-bit variable to store the result in.
```
Note: `batt_curr[out]`

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+/**
+ * @brief HK ADC read CITIROC temperature (deg. C) function
+ *
+ * This function returns the raw voltage value (Vtemp) as read by the ADC,
```
I would rephrase "returns" here, to avoid confusion in the future.

sonals301 replied:
Same as above

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+ * equivalent to CITIROC ASIC temperature value. Conversion to actual temperature
+ * value is performed by the ground software (CITIROC UI tool).
+ *
+ * @param citi_temp[in]     pointer to unsigned 16-bit variable to store the result in.
```
Note: `citi_temp[out]`

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+
+    if (err == HK_ADC_NO_ERR)
+    {
+        // Let the rest of the configurations remain intact, just change MUX input config and start single-shot conversion
```
80 cols

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+
+    if (err == HK_ADC_NO_ERR)
+    {
+        // Let the rest of the configurations remain intact, just change MUX input config and start single-shot conversion
```
80 cols

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+            err = hk_adc_reg_read(HK_ADC_REG_CONVERSION, &read_value);
+            if (err == HK_ADC_NO_ERR)
+            {
+                // Right shift the received result by 4 bits. Conversion value are 12 most significant bits
```
80 cols

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+            err = hk_adc_reg_read(HK_ADC_REG_CONVERSION, &read_value);
+            if (err == HK_ADC_NO_ERR)
+            {
+                // Right shift the received result by 4 bits. Conversion value are 12 most significant bits
```
80 cols

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+    send_buffer[2] = *(write_buffer + 1);   //LSB
+
+    // Write data to slave.
+    MSS_I2C_write(&g_mss_i2c0, HK_ADC_ADDRESS_GND, &send_buffer[0], sizeof(send_buffer)/sizeof(send_buffer[0]), MSS_I2C_RELEASE_BUS);
```
80 cols

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+    }
+
+    // Write to Address Pointer Register - pointing to 'reg'
+    MSS_I2C_write(&g_mss_i2c0, HK_ADC_ADDRESS_GND, &send_buffer, TX_LENGTH, MSS_I2C_RELEASE_BUS);
```
80 cols

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+    if (status == MSS_I2C_SUCCESS)
+    {
+        // Read data from target slave (we always read 2B)
+        MSS_I2C_read(&g_mss_i2c0, HK_ADC_ADDRESS_GND, &rx_buffer[0], RX_LENGTH, MSS_I2C_RELEASE_BUS);
```
80 cols

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+/* Local Function prototypes */
+static int hk_adc_reg_write(hk_adc_register_t reg, uint8_t *write_buffer);
+static int hk_adc_reg_read(hk_adc_register_t reg, uint16_t *read_buffer);
+static void hk_adc_update_multiplier_volt(void);
```
Function no longer exists, please remove the prototype declaration.

sonals301 replied:
Sorry, missed out on this one previously. Removed it now.

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+ * @return Average battery voltage raw value.
+ *
+ */
+uint16_t hk_adc_calc_avg_voltage(void)
```
I see no particular point in having the local `calc` functions. Might as well just move the contents to the `get_` functions, or am I missing something?

sonals301 replied:
Nothing big, just that the `get_` fns act as exported interface fns to get the calculated avg values. I have removed this and exposed the `calc_` fns instead. Hope it is okay now.

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+ */
+uint16_t hk_adc_calc_avg_voltage(void)
+{
+    uint16_t sum = 0;
```
if  `NUM_SAMPLES_TO_AVG` at some point becomes "16", we run the risk of overflow here... Please see my following comments for suggestions of improvement.

sonals301 replied:
Done

sonals301 replied:
Made the change as per other comments.

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+{
+    uint16_t sum = 0;
+    uint16_t average = 0;
+    for (int i = 0; i < NUM_SAMPLES_TO_AVG; i ++)
```
Suggest we "hard-code" the number of samples to average here, to perhaps 8?

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+    {
+        hk_adc_conv_read_volt(&volt_array[i]);
+        sum += volt_array[i];
+        average = (i == NUM_SAMPLES_TO_AVG - 1) ? sum / (uint16_t)NUM_SAMPLES_TO_AVG : 0;
```
This seems a bit clunky here... If you set the number of samples to average to **8**, then you could move this statement outside of the for loop and do `average = sum >> 3`, or maybe even `return (sum >> 3)`, avoiding the extra variable?

sonals301 replied:
Okay, sounds good to me. Made the change

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+    uint16_t average = 0;
+    for (int i = 0; i < NUM_SAMPLES_TO_AVG; i ++)
+    {
+        hk_adc_conv_read_volt(&volt_array[i]);
```
I would define `volt_array` inside the function, or was there any other reason you defined it as `static`? I see no other use of the variable outside of this function...

sonals301 replied:
True, I made it local now

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+ * @return Average battery current raw value.
+ *
+ */
+uint16_t hk_adc_calc_avg_current(void)
```
Same comments to this function as to `hk_adc_calc_avg_volt`

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,530 @@
+ * @return Average CITIROC temperature raw value.
+ *
+ */
+uint16_t hk_adc_calc_avg_citi_temp(void)
```
Same comments as to `hk_adc_calc_avg_volt()`

sonals301 replied:
Done

tstana commented on hk_adc/hk_adc.h:
```diff
@@ -0,0 +1,133 @@
+/*
```
GPL header. Never mind all the lines > 80 cols, in the interest of saving time, just leave them as they are!

sonals301 replied:
I made the changes :) It did not take much time

tstana replied:
Thank you, much appreciated -- and it looks much better now! 🙂

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -726,7 +735,7 @@ void msp_exprecv_syscommand(unsigned char opcode)
 {
 	citiroc_daq_set_hvps_temp(hvps_get_temp());
-	// citiroc_daq_set_citi_temp(hkadc_read(CITI_TEMP_CHAN));
+    citiroc_daq_set_citi_temp(hk_adc_get_avg_citi_temp());
```
Please use a TAB character here.

sonals301 replied:
I could not see this in my Eclipse IDE. I changed the settings and now I can see tabs and space characters. Made the change.

tstana commented on msp/msp_exp_handler.c:
```diff
@@ -413,8 +417,13 @@ void msp_expsend_start(unsigned char opcode, unsigned long *len)
 		send_data_hk[37] = (unsigned char) (count >> 0)  & 0xff;
 
+        /* ADC HK */
+        sprintf(((char*)send_data_hk)+38, "%04u", hk_adc_get_avg_volt());
+        sprintf(((char*)send_data_hk)+42, "%04u", hk_adc_get_avg_curr());
+        sprintf(((char*)send_data_hk)+46, "%04u", hk_adc_get_avg_citi_temp());
```
Please use TAB characters here...

sonals301 replied:
Done

tstana reviewed (changes requested):
Hi Sonal, I went through your code and made some comments. Hope they won't demoralize you too much, most of them are asthaetic and pretty straightforward to go through. Great job on the functions overall, amazing that they are working!

Commit:
addressed MR review comments

sonals301 commented:
Hi Thedi,
I addressed all your comments. Please pardon me for so many silly mistakes I made. I hope it is all better now. Let me know if there is anything else.

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,504 @@
+ * @return HK_ADC_ERR_READ_FAILED if any error occurs otherwise HK_ADC_NO_ERR.
+ *
+ */
+int hk_adc_reg_read(hk_adc_register_t reg, uint16_t *read_buffer)
```
Missing a `static` here (for clarity, it would probably not be an issue globally).

sonals301 replied:
Yeah, sorry. Added now.

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,504 @@
+ * @return HK_ADC_ERR_WRITE_FAILED if any error occurs otherwise HK_ADC_NO_ERR.
+ *
+ */
+int hk_adc_reg_write(hk_adc_register_t reg, uint8_t *write_buffer)
```
Missing a `static` here (for clarity, it would probably not be an issue globally).

sonals301 replied:
Added now.

tstana commented on hk_adc/hk_adc.c:
```diff
@@ -0,0 +1,504 @@
+static mss_i2c_status_t status;
+static uint16_t read_value;
+static uint8_t os_bit = 0;
+static uint8_t  rx_buffer[RX_LENGTH];
+static uint16_t hk_adc_avg_volt;
+static uint16_t hk_adc_avg_curr;
+static uint16_t hk_adc_avg_citi_temp;
```
These var's are all only used within one function, so they can be made local to these functions. It would also feel safer to me if the variables above them would also be made local to the functions they are used in, there is less risk that if one function gets changed at a later point, it breaks another's assignment to the variable. But if you have no time to address the variables above, then just open up an issue on it and let it be.

sonals301 replied:
I made these variables local. But I think the `_avg_` variables should be left global because they are returned outside this file when these fns are called in `msp_exphandler.c`.

tstana replied:
No, they are not needed. When you return from the `_calc_avg_` functions, they are effectively deallocated and the return value stored in a processor register. When the functions are called from `msp_exp_handler.c`, the call allocates the return value from a processor register to the `sprintf` parameter.

But no need to worry about it too much, it is just a minor nice-to-have detail... I'll just open an issue on this and mark this conversation as resolved.

tstana reviewed (changes requested):
No worries, they weren't silly mistakes at all! I found some more light issues to be addressed, hopefully they won't take too long.

Commit:
working codegit add . plus some minor changes

Commit:
removed commented code

Commit:
addressed MR comments
